src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -235,12 +232,8 @@ class TaskReplaceActor(
                 logger.info(s"Killing old ${nextOldInstance.instanceId}")
             }
 
-            if (runSpec.isResident) {
-              await(instanceTracker.setGoal(nextOldInstance.instanceId, Goal.Stopped))
-            } else {
-              await(instanceTracker.setGoal(nextOldInstance.instanceId, Goal.Decommissioned))
-            }
-            await(killService.killInstance(nextOldInstance, KillReason.Upgrading))
+            val goal = if (runSpec.isResident) Goal.Stopped else Goal.Decommissioned
+            await(instanceTracker.setGoal(nextOldInstance.instanceId, goal, GoalChangeReason.Upgrading))
```

Comment:
I like this. However, we have to be aware that this future is done once the goal is not but not when the instance is killed. This should be fine if the loop is stable.

Reply:
yeah I think that is fine, many places don't wait on things being killed anyway